inloop/statistics/views.py

```diff
@@ -0,0 +1,137 @@
+    status_code = 400
+
+
+class AdminView(UserPassesTestMixin, LoginRequiredMixin, View):
+    def test_func(self):
+        if not self.request.user:
+            return False
+        return self.request.user.is_superuser or self.request.user.is_staff
+
+
+class StatisticsView(AdminView, TemplateView):
+    template_name = "statistics/index.html"
+
+
+class SolutionsHistogramView(AdminView, TemplateView):
+    template_name = "statistics/solutions_histogram.html"
+
+    def get_context_data(self):
+        return {
+            "categories": Category.objects.all()
+        }
+
+    def post(self, request) -> JsonResponse:
+        try:
+            data = json.loads(request.body)
+        except JSONDecodeError:
+            return MalformedJson()
+
+        from_timestamp = data.get("from_timestamp", None)
+        to_timestamp = data.get("to_timestamp", None)
+        passed = data.get("passed", None)
+        category_id = data.get("category_id", None)
+        granularity = data.get("granularity", "minute")
+
+        try:
+            truncator = get_truncator(field_name="submission_date", granularity=granularity)
+        except KeyError:
+            return MalformedJson()
+
+        queryset = Solution.objects
+        if from_timestamp:
+            queryset = queryset.filter(submission_date__gte=from_timestamp)
+        if to_timestamp:
+            queryset = queryset.filter(submission_date__lte=to_timestamp)
+        if passed is not None:
+            queryset = queryset.filter(passed=passed)
+        if category_id:
+            queryset = queryset.filter(task__category_id=category_id)
+
+        histogram_queryset = queryset \
+            .annotate(date=truncator) \
+            .values("date") \
+            .annotate(count=models.Count("pk"))
+
+        return JsonResponse({
+            "histogram": list(histogram_queryset)
+        })
+
+
+class AttemptsHistogramView(AdminView, TemplateView):
+    template_name = "statistics/attempts_histogram.html"
+
+    def get_context_data(self):
+        return {
+            "tasks": Task.objects.all()
+        }
+
+    def post(self, request) -> JsonResponse:
+        try:
+            data = json.loads(request.body)
+        except JSONDecodeError:
+            return MalformedJson()
+
+        task_id = data.get("task_id")
+        if not task_id:
+            return MalformedJson()
+
+        # TODO: Check back later, if this computation is correct
+        queryset = Solution.objects\
+            .filter(passed=True, task_id=task_id)\
+            .annotate(tries_until_pass=models.Min("scoped_id"))\
+            .values("tries_until_pass")\
+            .order_by("-tries_until_pass")
+
+        # TODO: Group by on the queryset
+        grouped_entries = defaultdict(int)
+        for entry in queryset:
+            grouped_entries[entry["tries_until_pass"]] += 1
+
+        return JsonResponse({
+            "chart_data": list(grouped_entries.items())
+        })
```

Comment:
This works as intended :+1: I am thinking about renaming `scoped_id` to `attempt_no` or something similar because of that.

Suggestions:
- `num_trials`/`num_attempts` instead of `tries_until_pass`
- for the counting, you can use `collections.Counter` together with `qs.values_list(…, flat=True)`
- double check if `group_by(…)` and `list(grouped_entries.items())` is really necessary (does the JS code depend on that?)

```python
from collections import Counter
queryset = Solution.objects.filter(
    passed=True,
    task_id=task_id
).annotate(
    num_trials=Min('scoped_id')
)
histogram = Counter(queryset.values_list('num_trials', flat=True))
return JsonResponse(histogram)
```

Will give a JSON document à la

```json
{"1": 10, "2": 8, "3": 11, "4": 11, "5": 7, "7": 8, "8": 8, "6": 9, "11": 6, "12": 1, "13": 2, "10": 3, "9": 6, "15": 2, "16": 1, "17": 2, "18": 1, "14": 1}
```